store/postgres/src/relational_queries.rs

```diff
@@ -976,13 +1009,52 @@ impl<'a> FilterQuery<'a> {
     }
 
     fn limit(&self, out: &mut AstPass<Pg>) {
-        if let Some(first) = &self.first {
+        if let Some(first) = &self.range.first {
             out.push_sql("\n limit ");
-            out.push_sql(first);
+            out.push_sql(&first.to_string());
         }
-        if let Some(skip) = &self.skip {
+        if self.range.skip > 0 {
             out.push_sql("\noffset ");
-            out.push_sql(skip);
+            out.push_sql(&self.range.skip.to_string());
+        }
+    }
+
+    fn rank_window(&self, out: &mut AstPass<Pg>) -> QueryResult<()> {
+        if let Some(window) = self.window {
+            out.push_sql(", rank() over (partition by ");
+            out.push_identifier(window.as_str())?;
+            out.push_sql(" ");
+            self.order_by(out)?;
+            out.push_sql(") as pos");
+        }
+        Ok(())
+    }
+
+    // Generate (taking optionality of either clause into account)
+    //    where {pos} > {order.skip} and {pos} <= {order.first + order.skip}
+    //  `pos` must be the name of the column we order by and 'conj' must be
+    // either 'and ' or 'where ' depending on where we add this clause
+    fn limit_per_window(&self, pos: &str, conj: &str, out: &mut AstPass<Pg>) {
+        if self.window.is_some() {
+            let have_first = self.range.first.is_some();
+            let have_skip = self.range.skip > 0;
+            if have_first || have_skip {
+                out.push_sql("\n ");
+                out.push_sql(conj);
+            }
+            if have_skip {
+                out.push_sql(pos);
+                out.push_sql(" > ");
+                out.push_sql(&self.range.skip.to_string());
+                if self.range.first.is_some() {
+                    out.push_sql(" and ");
+                }
+            }
+            if let Some(first) = self.range.first {
+                out.push_sql(pos);
+                out.push_sql(" <= ");
+                out.push_sql(&(first + self.range.skip).to_string());
+            }
```

Comment:
I can't quite follow the SQL generation in this file (the highlighted code is just one example). Would it make sense to have some tests for the SQL generation that make it easier to compare the logic against examples?

Reply:
Putting SQL directly in tests is very brittle. The relational schema generation has some of that, and it's a frustrating exercise in counting spaces every time something changes there.

What I usually do to look at the SQL is to set `log_statement = 'all'` in my `postgresql.conf`, run a test that exercises what I am interested in, and then look at the Postgres logs to see the SQL